File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.java
/*
 * Copyright (C) 2011 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package okhttp3.mockwebserver;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;
import okhttp3.Headers;
import okhttp3.internal.Internal;
import okhttp3.internal.http2.Settings;
import okhttp3.WebSocketListener;
import okio.Buffer;

/** A scripted response to be replayed by the mock web server. */
public final class MockResponse implements Cloneable {
  private static final String CHUNKED_BODY_HEADER = "Transfer-encoding: chunked";

  private String status;
  private Headers.Builder headers = new Headers.Builder();

  private Buffer body;

  private long throttleBytesPerPeriod = Long.MAX_VALUE;
  private long throttlePeriodAmount = 1;
  private TimeUnit throttlePeriodUnit = TimeUnit.SECONDS;

  private SocketPolicy socketPolicy = SocketPolicy.KEEP_OPEN;
  private int http2ErrorCode = -1;

  private long bodyDelayAmount = 0;
  private TimeUnit bodyDelayUnit = TimeUnit.MILLISECONDS;

  private long shutdownDelay = 0;
  private TimeUnit shutdownDelayUnit = TimeUnit.MILLISECONDS;

  private List<PushPromise> promises = new ArrayList<>();
  private Settings settings;
  private WebSocketListener webSocketListener;

  /** Creates a new mock response with an empty body. */
  public MockResponse() {
    setResponseCode(200);
    setHeader("Content-Length", 0);
  }

  @Override public MockResponse clone() {
    try {
      MockResponse result = (MockResponse) super.clone();
      result.headers = headers.build().newBuilder();
      result.promises = new ArrayList<>(promises);
      return result;
    } catch (CloneNotSupportedException e) {
      throw new AssertionError();
    }
  }

  /** Returns the HTTP response line, such as "HTTP/1.1 200 OK". */
  public String getStatus() {
    return status;
  }

  public MockResponse setResponseCode(int code) {
    String reason = "Mock Response";
    if (code >= 100 && code < 200) {
      reason = "Informational";
    } else if (code >= 200 && code < 300) {
      reason = "OK";
    } else if (code >= 300 && code < 400) {
      reason = "Redirection";
    } else if (code >= 400 && code < 500) {
      reason = "Client Error";
    } else if (code >= 500 && code < 600) {
      reason = "Server Error";
    }
    return setStatus("HTTP/1.1 " + code + " " + reason);
  }

  public MockResponse setStatus(String status) {
    this.status = status;
    return this;
  }

  /** Returns the HTTP headers, such as "Content-Length: 0". */
  public Headers getHeaders() {
    return headers.build();
  }

  /**
   * Removes all HTTP headers including any "Content-Length" and "Transfer-encoding" headers that
   * were added by default.
   */
  public MockResponse clearHeaders() {
    headers = new Headers.Builder();
    return this;
  }

  /**
   * Adds {@code header} as an HTTP header. For well-formed HTTP {@code header} should contain a
   * name followed by a colon and a value.
   */
  public MockResponse addHeader(String header) {
    headers.add(header);
    return this;
  }

  /**
   * Adds a new header with the name and value. This may be used to add multiple headers with the
   * same name.
   */
  public MockResponse addHeader(String name, Object value) {
    headers.add(name, String.valueOf(value));
    return this;
  }

  /**
   * Adds a new header with the name and value. This may be used to add multiple headers with the
   * same name. Unlike {@link #addHeader(String, Object)} this does not validate the name and
   * value.
   */
  public MockResponse addHeaderLenient(String name, Object value) {
    Internal.instance.addLenient(headers, name, String.valueOf(value));
    return this;
  }

  /**
   * Removes all headers named {@code name}, then adds a new header with the name and value.
   */
  public MockResponse setHeader(String name, Object value) {
    removeHeader(name);
    return addHeader(name, value);
  }

  /** Replaces all headers with those specified in {@code headers}. */
  public MockResponse setHeaders(Headers headers) {
    this.headers = headers.newBuilder();
    return this;
  }

  /** Removes all headers named {@code name}. */
  public MockResponse removeHeader(String name) {
    headers.removeAll(name);
    return this;
  }

  /** Returns a copy of the raw HTTP payload. */
  public Buffer getBody() {
    return body != null ? body.clone() : null;
  }

  public MockResponse setBody(Buffer body) {
    setHeader("Content-Length", body.size());
    this.body = body.clone(); // Defensive copy.
    return this;
  }

  /** Sets the response body to the UTF-8 encoded bytes of {@code body}. */
  public MockResponse setBody(String body) {
    return setBody(new Buffer().writeUtf8(body));
  }

  /**
   * Sets the response body to {@code body}, chunked every {@code maxChunkSize} bytes.
   */
  public MockResponse setChunkedBody(Buffer body, int maxChunkSize) {
    removeHeader("Content-Length");
    headers.add(CHUNKED_BODY_HEADER);

    Buffer bytesOut = new Buffer();
    while (!body.exhausted()) {
      long chunkSize = Math.min(body.size(), maxChunkSize);
      bytesOut.writeHexadecimalUnsignedLong(chunkSize);
      bytesOut.writeUtf8("\r\n");
      bytesOut.write(body, chunkSize);
      bytesOut.writeUtf8("\r\n");
    }
    bytesOut.writeUtf8("0\r\n\r\n"); // Last chunk + empty trailer + CRLF.

    this.body = bytesOut;
    return this;
  }

  /**
   * Sets the response body to the UTF-8 encoded bytes of {@code body}, chunked every {@code
   * maxChunkSize} bytes.
   */
  public MockResponse setChunkedBody(String body, int maxChunkSize) {
    return setChunkedBody(new Buffer().writeUtf8(body), maxChunkSize);
  }

  public SocketPolicy getSocketPolicy() {
    return socketPolicy;
  }

  public MockResponse setSocketPolicy(SocketPolicy socketPolicy) {
    this.socketPolicy = socketPolicy;
    return this;
  }

  public int getHttp2ErrorCode() {
    return http2ErrorCode;
  }

  /**
   * Sets the <a href="https://tools.ietf.org/html/rfc7540#section-7">HTTP/2 error code</a> to be
   * returned when resetting the stream. This is only valid with {@link
   * SocketPolicy#RESET_STREAM_AT_START}.
   */
  public MockResponse setHttp2ErrorCode(int http2ErrorCode) {
    this.http2ErrorCode = http2ErrorCode;
    return this;
  }

  /**
   * Throttles the request reader and response writer to sleep for the given period after each
   * series of {@code bytesPerPeriod} bytes are transferred. Use this to simulate network behavior.
   */
  public MockResponse throttleBody(long bytesPerPeriod, long period, TimeUnit unit) {
    this.throttleBytesPerPeriod = bytesPerPeriod;
    this.throttlePeriodAmount = period;
    this.throttlePeriodUnit = unit;
    return this;
  }

  public long getThrottleBytesPerPeriod() {
    return throttleBytesPerPeriod;
  }

  public long getThrottlePeriod(TimeUnit unit) {
    return unit.convert(throttlePeriodAmount, throttlePeriodUnit);
  }

  /**
   * Set the delayed time of the response body to {@code delay}. This applies to the response body
   * only; response headers are not affected.
   */
  public MockResponse setBodyDelay(long delay, TimeUnit unit) {
    bodyDelayAmount = delay;
    bodyDelayUnit = unit;
    return this;
  }

  public long getBodyDelay(TimeUnit unit) {
    return unit.convert(bodyDelayAmount, bodyDelayUnit);
  }

  /**
   * Set the time to {@code delay} prior to shutting down an HTTP/2 connection. This is only valid
   * with {@link SocketPolicy#DISCONNECT_AT_END}. Use this to simulate a graceful HTTP/2
   * connection shutdown.
   */
  public MockResponse setShutdownDelay(long delay, TimeUnit unit) {
    shutdownDelay = delay;
    shutdownDelayUnit = unit;
    return this;
  }

  public long getShutdownDelay(TimeUnit unit) {
    return unit.convert(shutdownDelay, shutdownDelayUnit);
  }

  /**
   * When {@link MockWebServer#setProtocols(java.util.List) protocols} include {@linkplain
   * okhttp3.Protocol#HTTP_2}, this attaches a pushed stream to this response.
   */
  public MockResponse withPush(PushPromise promise) {
    this.promises.add(promise);
    return this;
  }

  /** Returns the streams the server will push with this response. */
  public List<PushPromise> getPushPromises() {
    return promises;
  }

  /**
   * When {@linkplain MockWebServer#setProtocols(java.util.List) protocols} include {@linkplain
   * okhttp3.Protocol#HTTP_2 HTTP/2}, this pushes {@code settings} before writing the response.
   */
  public MockResponse withSettings(Settings settings) {
    this.settings = settings;
    return this;
  }

  public Settings getSettings() {
    return settings;
  }

  /**
   * Attempts to perform a web socket upgrade on the connection. This will overwrite any previously
   * set status or body.
   */
  public MockResponse withWebSocketUpgrade(WebSocketListener listener) {
    setStatus("HTTP/1.1 101 Switching Protocols");
    setHeader("Connection", "Upgrade");
    setHeader("Upgrade", "websocket");
    body = null;
    webSocketListener = listener;
    return this;
  }

  public WebSocketListener getWebSocketListener() {
    return webSocketListener;
  }

  @Override public String toString() {
    return status;
  }
}

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt
/*
 * Copyright (C) 2011 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package okhttp3.mockwebserver

import okhttp3.Headers
import okhttp3.WebSocketListener
import okhttp3.internal.addHeaderLenient
import okhttp3.internal.http2.Settings
import okhttp3.mockwebserver.internal.duplex.DuplexResponseBody
import okio.Buffer
import java.util.concurrent.TimeUnit

/** A scripted response to be replayed by the mock web server. */
class MockResponse : Cloneable {
  /** Returns the HTTP response line, such as "HTTP/1.1 200 OK". */
  @set:JvmName("status")
  var status: String = ""

  private var headersBuilder = Headers.Builder()
  private var trailersBuilder = Headers.Builder()

  /** The HTTP headers, such as "Content-Length: 0". */
  @set:JvmName("headers")
  var headers: Headers
    get() = headersBuilder.build()
    set(value) {
      this.headersBuilder = value.newBuilder()
    }

  @set:JvmName("trailers")
  var trailers: Headers
    get() = trailersBuilder.build()
    set(value) {
      this.trailersBuilder = value.newBuilder()
    }

  private var body: Buffer? = null

  var throttleBytesPerPeriod = Long.MAX_VALUE
    private set
  private var throttlePeriodAmount = 1L
  private var throttlePeriodUnit = TimeUnit.SECONDS

  @set:JvmName("socketPolicy")
  var socketPolicy = SocketPolicy.KEEP_OPEN

  /**
   * Sets the [HTTP/2 error code](https://tools.ietf.org/html/rfc7540#section-7) to be
   * returned when resetting the stream.
   * This is only valid with [SocketPolicy.RESET_STREAM_AT_START].
   */
  @set:JvmName("http2ErrorCode")
  var http2ErrorCode = -1

  private var bodyDelayAmount = 0L
  private var bodyDelayUnit = TimeUnit.MILLISECONDS

  private var headersDelayAmount = 0L
  private var headersDelayUnit = TimeUnit.MILLISECONDS

  private var promises = mutableListOf<PushPromise>()
  var settings: Settings = Settings()
    private set
  var webSocketListener: WebSocketListener? = null
    private set
  var duplexResponseBody: DuplexResponseBody? = null
    private set
  val isDuplex: Boolean
    get() = duplexResponseBody != null

  /** Returns the streams the server will push with this response. */
  val pushPromises: List<PushPromise>
    get() = promises

  /** Creates a new mock response with an empty body. */
  init {
    setResponseCode(200)
    setHeader("Content-Length", 0L)
  }

  public override fun clone(): MockResponse {
    val result = super.clone() as MockResponse
    result.headersBuilder = headersBuilder.build().newBuilder()
    result.promises = promises.toMutableList()
    return result
  }

  @JvmName("-deprecated_getStatus")
  @Deprecated(
      message = "moved to var",
      replaceWith = ReplaceWith(expression = "status"),
      level = DeprecationLevel.ERROR)
  fun getStatus(): String = status

  @Deprecated(
      message = "moved to var.  Replace setStatus(...) with status(...) to fix Java",
      replaceWith = ReplaceWith(expression = "apply { this.status = status }"),
      level = DeprecationLevel.WARNING)
  fun setStatus(status: String) = apply {
    this.status = status
  }

  fun setResponseCode(code: Int): MockResponse {
    val reason = when (code) {
      in 100..199 -> "Informational"
      in 200..299 -> "OK"
      in 300..399 -> "Redirection"
      in 400..499 -> "Client Error"
      in 500..599 -> "Server Error"
      else -> "Mock Response"
    }
    return apply { status = "HTTP/1.1 $code $reason" }
  }

  /**
   * Removes all HTTP headers including any "Content-Length" and "Transfer-encoding" headers that
   * were added by default.
   */
  fun clearHeaders() = apply {
    headersBuilder = Headers.Builder()
  }

  /**
   * Adds [header] as an HTTP header. For well-formed HTTP [header] should contain a
   * name followed by a colon and a value.
   */
  fun addHeader(header: String) = apply {
    headersBuilder.add(header)
  }

  /**
   * Adds a new header with the name and value. This may be used to add multiple headers with the
   * same name.
   */
  fun addHeader(name: String, value: Any) = apply {
    headersBuilder.add(name, value.toString())
  }

  /**
   * Adds a new header with the name and value. This may be used to add multiple headers with the
   * same name. Unlike [addHeader] this does not validate the name and
   * value.
   */
  fun addHeaderLenient(name: String, value: Any) = apply {
    addHeaderLenient(headersBuilder, name, value.toString())
  }

  /**
   * Removes all headers named [name], then adds a new header with the name and value.
   */
  fun setHeader(name: String, value: Any) = apply {
    removeHeader(name)
    addHeader(name, value)
  }

  /** Removes all headers named [name]. */
  fun removeHeader(name: String) = apply {
    headersBuilder.removeAll(name)
  }

  /** Returns a copy of the raw HTTP payload. */
  fun getBody(): Buffer? = body?.clone()

  fun setBody(body: Buffer) = apply {
    setHeader("Content-Length", body.size)
    this.body = body.clone() // Defensive copy.
  }

  /** Sets the response body to the UTF-8 encoded bytes of [body]. */
  fun setBody(body: String): MockResponse = setBody(Buffer().writeUtf8(body))

  fun setBody(duplexResponseBody: DuplexResponseBody) = apply {
    this.duplexResponseBody = duplexResponseBody
  }

  /**
   * Sets the response body to [body], chunked every [maxChunkSize] bytes.
   */
  fun setChunkedBody(body: Buffer, maxChunkSize: Int) = apply {
    removeHeader("Content-Length")
    headersBuilder.add(CHUNKED_BODY_HEADER)

    val bytesOut = Buffer()
    while (!body.exhausted()) {
      val chunkSize = minOf(body.size, maxChunkSize.toLong())
      bytesOut.writeHexadecimalUnsignedLong(chunkSize)
      bytesOut.writeUtf8("\r\n")
      bytesOut.write(body, chunkSize)
      bytesOut.writeUtf8("\r\n")
    }
    bytesOut.writeUtf8("0\r\n") // Last chunk. Trailers follow!
    this.body = bytesOut
  }

  /**
   * Sets the response body to the UTF-8 encoded bytes of [body],
   * chunked every [maxChunkSize] bytes.
   */
  fun setChunkedBody(body: String, maxChunkSize: Int): MockResponse =
    setChunkedBody(Buffer().writeUtf8(body), maxChunkSize)

  @JvmName("-deprecated_getHeaders")
  @Deprecated(
      message = "moved to var",
      replaceWith = ReplaceWith(expression = "headers"),
      level = DeprecationLevel.ERROR)
  fun getHeaders(): Headers = headers

  @Deprecated(
      message = "moved to var. Replace setHeaders(...) with headers(...) to fix Java",
      replaceWith = ReplaceWith(expression = "apply { this.headers = headers }"),
      level = DeprecationLevel.WARNING)
  fun setHeaders(headers: Headers) = apply { this.headers = headers }

  @JvmName("-deprecated_getTrailers")
  @Deprecated(
      message = "moved to var",
      replaceWith = ReplaceWith(expression = "trailers"),
      level = DeprecationLevel.ERROR)
  fun getTrailers(): Headers = trailers

  @Deprecated(
      message = "moved to var. Replace setTrailers(...) with trailers(...) to fix Java",
      replaceWith = ReplaceWith(expression = "apply { this.trailers = trailers }"),
      level = DeprecationLevel.WARNING)
  fun setTrailers(trailers: Headers) = apply { this.trailers = trailers }

  @JvmName("-deprecated_getSocketPolicy")
  @Deprecated(
      message = "moved to var",
      replaceWith = ReplaceWith(expression = "socketPolicy"),
      level = DeprecationLevel.ERROR)
  fun getSocketPolicy() = socketPolicy

  @Deprecated(
      message = "moved to var. Replace setSocketPolicy(...) with socketPolicy(...) to fix Java",
      replaceWith = ReplaceWith(expression = "apply { this.socketPolicy = socketPolicy }"),
      level = DeprecationLevel.WARNING)
  fun setSocketPolicy(socketPolicy: SocketPolicy) = apply {
    this.socketPolicy = socketPolicy
  }

  @JvmName("-deprecated_getHttp2ErrorCode")
  @Deprecated(
      message = "moved to var",
      replaceWith = ReplaceWith(expression = "http2ErrorCode"),
      level = DeprecationLevel.ERROR)
  fun getHttp2ErrorCode() = http2ErrorCode

  @Deprecated(
      message = "moved to var. Replace setHttp2ErrorCode(...) with http2ErrorCode(...) to fix Java",
      replaceWith = ReplaceWith(expression = "apply { this.http2ErrorCode = http2ErrorCode }"),
      level = DeprecationLevel.WARNING)
  fun setHttp2ErrorCode(http2ErrorCode: Int) = apply {
    this.http2ErrorCode = http2ErrorCode
  }

  /**
   * Throttles the request reader and response writer to sleep for the given period after each
   * series of [bytesPerPeriod] bytes are transferred. Use this to simulate network behavior.
   */
  fun throttleBody(bytesPerPeriod: Long, period: Long, unit: TimeUnit) = apply {
    throttleBytesPerPeriod = bytesPerPeriod
    throttlePeriodAmount = period
    throttlePeriodUnit = unit
  }

  fun getThrottlePeriod(unit: TimeUnit): Long =
    unit.convert(throttlePeriodAmount, throttlePeriodUnit)

  /**
   * Set the delayed time of the response body to [delay]. This applies to the response body
   * only; response headers are not affected.
   */
  fun setBodyDelay(delay: Long, unit: TimeUnit) = apply {
    bodyDelayAmount = delay
    bodyDelayUnit = unit
  }

  fun getBodyDelay(unit: TimeUnit): Long =
    unit.convert(bodyDelayAmount, bodyDelayUnit)

  fun setHeadersDelay(delay: Long, unit: TimeUnit) = apply {
    headersDelayAmount = delay
    headersDelayUnit = unit
  }

  fun getHeadersDelay(unit: TimeUnit): Long =
    unit.convert(headersDelayAmount, headersDelayUnit)

  /**
   * When [protocols][MockWebServer.protocols] include [HTTP_2][okhttp3.Protocol], this attaches a
   * pushed stream to this response.
   */
  fun withPush(promise: PushPromise) = apply {
    promises.add(promise)
  }

  /**
   * When [protocols][MockWebServer.protocols] include [HTTP_2][okhttp3.Protocol], this pushes
   * [settings] before writing the response.
   */
  fun withSettings(settings: Settings) = apply {
    this.settings = settings
  }

  /**
   * Attempts to perform a web socket upgrade on the connection.
   * This will overwrite any previously set status or body.
   */
  fun withWebSocketUpgrade(listener: WebSocketListener) = apply {
    status = "HTTP/1.1 101 Switching Protocols"
    setHeader("Connection", "Upgrade")
    setHeader("Upgrade", "websocket")
    body = null
    webSocketListener = listener
  }

  override fun toString() = status

  companion object {
    private const val CHUNKED_BODY_HEADER = "Transfer-encoding: chunked"
  }
}
